docs/backend-system/building-plugins-and-modules/02-testing.md

```diff
@@ -6,21 +6,162 @@ sidebar_label: Testing
+import { rest } from 'msw';
+import { setupServer } from 'msw/node';
+
+describe('read from remote', () => {
+  const worker = setupServer();
+  setupRequestMockHandlers(worker);
+
+  it('should auth and read successfully', async () => {
+    expect.assertions(1);
+
+    worker.use(
+      rest.get('https://remote-server.com/api/v3/foo', (req, res, ctx) => {
+        expect(req.headers.get('authorization')).toBe('Bearer fake');
+        return res(
+          ctx.status(200),
+          ctx.set('Content-Type', 'application/json'),
+          ctx.body(JSON.stringify({ value: 7 })),
+        );
+      }),
+    );
+
+    // exercise your plugin or service as usual, with real clients
   });
 });
 ```
+
+## Testing Database Interactions
+
+The `@backstage/backend-test-utils` package includes facilities for testing your
+plugins' interactions with databases, including spinning up `testcontainers`
+powered Docker images with real database engines to connect to.
+
+The base setup for such a test could look as follows:
+
+```ts
+// MyDatabaseClass.test.ts
+import { TestDatabaseId, TestDatabases } from '@backstage/backend-test-utils';
+import {
+  MyDatabaseClass,
+  applyDatabaseMigrations,
+  type FooTableRow,
+} from './MyDatabaseClass';
+
+describe('MyDatabaseClass', () => {
+  // Change this to the set of constants that you actually actively intend to
+  // support. Make sure to create only one TestDatabases instance per file,
+  // since spinning up "physical" databases to test against is much costlier
+  // than creating the "logical" databases within them that the individual
+  // tests use.
+  const databases = TestDatabases.create({
+    ids: ['POSTGRES_13', 'POSTGRES_9', 'SQLITE_3'],
+  });
+
+  // Just an example of how to conveniently bundle up the setup code
+  async function createSut(databaseId: TestDatabaseId) {
+    const knex = await databases.init(databaseId);
+    const sut = new MyDatabaseClass({ database: knex });
+    await sut.runMigrations();
+    return { knex, sut };
+  }
+
+  describe('foo', () => {
+    // Easily run the exact same test onto all supported databases
+    it.each(databases.eachSupportedId())(
+      'should run foo on %p',
+      async databaseId => {
+        const { knex, sut } = await createSut(databaseId);
+        // raw knex is available for underlying manipulation
+        await knex<FooTableRow>('foo').insert({ value: 2 });
+        // drive your system under test as usual
+        await expect(sut.foos()).resolves.toEqual([{ value: 2 }]);
+      });
+  });
+```
+
+If you want to pass the test database instance into backend plugins or services,
+you can supply it in the form of a mock instance of `coreServices.database` to
+your test database.
+
+```ts
+const { knex, sut } = await createSut(databaseId);
+const { server } = await startTestBackend({
+  features: [myPlugin()],
+  services: [[coreServices.database, { getClient: async () => knex }]],
```

Comment:
We should definitely have a utility for this :grin:

Reply:
Indeed. I'll try to get some time to clean that up